File: emulinker/src/main/java/org/emulinker/kaillera/controller/KailleraServerController.java
package org.emulinker.kaillera.controller;

import java.net.InetSocketAddress;
import org.emulinker.kaillera.model.KailleraServer;
import org.emulinker.kaillera.model.exception.*;

public interface KailleraServerController {
  public int newConnection(InetSocketAddress clientSocketAddress, String protocol)
      throws ServerFullException, NewConnectionException;

  public KailleraServer getServer();

  public int getBufferSize();

  public String getVersion();

  public int getNumClients();

  public String[] getClientTypes();

  public void start();

  public void stop();
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/KailleraServerController.kt
package org.emulinker.kaillera.controller

import java.net.InetSocketAddress
import org.emulinker.kaillera.controller.v086.V086ClientHandler
import org.emulinker.kaillera.model.KailleraServer
import org.emulinker.kaillera.model.exception.NewConnectionException
import org.emulinker.kaillera.model.exception.ServerFullException
import org.emulinker.net.UdpSocketProvider

interface KailleraServerController {
  val server: KailleraServer
  val bufferSize: Int
  val version: String
  val numClients: Int
  val clientTypes: Array<String>

  @Throws(ServerFullException::class, NewConnectionException::class)
  suspend fun newConnection(
      udpSocketProvider: UdpSocketProvider, clientSocketAddress: InetSocketAddress, protocol: String
  ): Int

  fun start()

  suspend fun stop()
  val clientHandlers: MutableMap<Int, V086ClientHandler>
}
